### PR TITLE
[REVIEW] Feature/concat columns refactor

### PR DESCRIPTION
**Pull request by williamBlazing**
This PR does lots of things... sorry.
It implements the join optimization for small tables
It refactors concat columns
If also allows for the normalization of data types during concat columns and also right before joins
This means you can now join a int32 against an int64 or do a union against similarly different types

### CODE REVIEW

**Comment by kharoc** [comment]
READY TO REVIEW!

1 RAL

*******************************************************************************
TOTAL SUMMARY for test suite: 
PASSED: 640/644
FAILED: 4/644
CRASH: 0/644
TOTAL: 644
*******************************************************************************
!!!!!!!!!!!!!!!! FAILED gdf TESTS !!!!!!!!!!!!!!!!
!!!!!!!!!!!!!!!! FAILED psv TESTS !!!!!!!!!!!!!!!!
!!!!!!!!!!!!!!!! FAILED parquet TESTS !!!!!!!!!!!!
!!!!!!!!!!!!!!!! FAILED orc TESTS !!!!!!!!!!!!!!!!
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Date Test - orc
    FAILED: 2/5 ['TEST_01', 'TEST_02']
    CRASH: 0/5
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Timestamp Test - orc
    FAILED: 1/12 ['TEST_02']
    CRASH: 0/12
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Cast Test - orc
    FAILED: 1/9 ['TEST_01']
    CRASH: 0/9
MAX DELTA: 248.0
*******************************************************************************


2 RALs


*******************************************************************************
TOTAL SUMMARY for test suite: 
PASSED: 466/472
FAILED: 6/472
CRASH: 0/472
TOTAL: 472
*******************************************************************************
!!!!!!!!!!!!!!!! FAILED gdf TESTS !!!!!!!!!!!!!!!!
!!!!!!!!!!!!!!!! FAILED psv TESTS !!!!!!!!!!!!!!!!
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Left outer join Test - psv
    FAILED: 1/4 ['TEST_01']
    CRASH: 0/4
!!!!!!!!!!!!!!!! FAILED parquet TESTS !!!!!!!!!!!!
!!!!!!!!!!!!!!!! FAILED orc TESTS !!!!!!!!!!!!!!!!
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Date Test - orc
    FAILED: 2/5 ['TEST_01', 'TEST_02']
    CRASH: 0/5
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Timestamp Test - orc
    FAILED: 1/12 ['TEST_02']
    CRASH: 0/12
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Left outer join Test - orc
    FAILED: 1/4 ['TEST_01']
    CRASH: 0/4
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Cast Test - orc
    FAILED: 1/9 ['TEST_01']
    CRASH: 0/9
MAX DELTA: 241.0
*******************************************************************************

File: src/distribution/primitives.cpp [hunk 1]
```diff
@@ -552,11 +524,30 @@ std::vector<NodeColumns> collectPartitions(const Context& context) {
...
+        if (received[node_idx]){
+          std::cout<<"ERROR: Already received collectSomePartitions from node: "<<node_idx<<std::endl;
+        } 
         node_columns.emplace_back(*node, column_message->getColumns());
+        received[node_idx] = true;
     }
     return node_columns;
 }
 
+void scatterData(const Context& context, std::vector<gdf_column_cpp>& table) {
```
Comment (felipeblazing) [inline src/distribution/primitives.cpp]: how is this related to concatting?

File: src/distribution/primitives.cpp [hunk 2]
```diff
@@ -27,7 +27,7 @@
 #include "copying.hpp"
 #include "cudf/merge.hpp"
 #include "cudf/search.hpp"
-
+#include "utilities/CommonOperations.h"
```
Comment (felipeblazing) [inline src/distribution/primitives.cpp]: how are the changes here part of concat?

File: src/distribution/primitives.cpp [hunk 3]
```diff
@@ -742,22 +736,98 @@ std::vector<gdf_size_type> collectRowSize(const Context& context) {
...
+    }
+    node_row_sizes[node_idx] = concrete_message->getTotalRowSize();    
+    received[node_idx] = true;
   }
 
   return node_row_sizes;
 }
 
+
+void distributeLeftRightNumRows(const Context& context, std::size_t left_num_rows, std::size_t right_num_rows) {
```
Comment (felipeblazing) [inline src/distribution/primitives.cpp]: how is this related to concat?

File: src/operators/JoinOperator.cpp [hunk 4]
```diff
@@ -66,6 +67,8 @@ class DistributedJoinOperator : public JoinOperator {
 protected:
     blazing_frame process_distribution(blazing_frame& frame, const std::string& query);
 
+    blazing_frame process_hash_based_distribution(blazing_frame& frame, const std::string& query);
```
Comment (felipeblazing) [inline src/operators/JoinOperator.cpp]: this seems not related to concat

File: src/operators/JoinOperator.cpp [hunk 5]
```diff
@@ -166,23 +169,28 @@ DistributedJoinOperator::DistributedJoinOperator(const Context* context)
 
 blazing_frame DistributedJoinOperator::operator()(blazing_frame& frame, const std::string& query) {
     // Execute distribution
-    auto distributed_frame = process_distribution(frame, query);
+    std::string join_type = get_named_expression(query, "joinType");
+    if(join_type == INNER_JOIN){
+        frame = process_distribution(frame, query);
```
Comment (felipeblazing) [inline src/operators/JoinOperator.cpp]: is this an optimizatino for when we dont need to hash a column? when is this used how is this used? how is this related?

File: src/operators/JoinOperator.cpp [hunk 6]
```diff
@@ -205,6 +213,106 @@ std::vector<gdf_column_cpp> DistributedJoinOperator::process_distribution_table(
 }
 
 blazing_frame DistributedJoinOperator::process_distribution(blazing_frame& frame, const std::string& query) {
+
```
Comment (felipeblazing) [inline src/operators/JoinOperator.cpp]: this seem slike changes for your optimized join whihc definietly has nothign to do with concat

File: src/CalciteExpressionParsing.cpp [hunk 7]
```diff
@@ -35,6 +35,13 @@ bool is_type_float(gdf_dtype type){
 			GDF_FLOAT64 == type);
 }
 
+bool is_type_decimal(gdf_dtype type){
```
[the remark this answers is not in the document]
Comment (williamBlazing) [reply src/CalciteExpressionParsing.cpp]: oh, yeah, it should be is_type_integer
i will change that

File: src/CalciteExpressionParsing.cpp [hunk 8]
```diff
@@ -56,24 +63,6 @@ bool is_numeric_type(gdf_dtype type){
 }
 
 gdf_dtype get_next_biggest_type(gdf_dtype type){
-	//	if(type == GDF_INT8){
```
[the remark this answers is not in the document]
Comment (williamBlazing) [reply src/CalciteExpressionParsing.cpp]: all that was deleted was commented out deprecated bad code

File: src/CalciteExpressionParsing.cpp [hunk 9]
```diff
@@ -291,6 +280,84 @@ gdf_dtype get_output_type(gdf_dtype input_left_type, gdf_dtype input_right_type,
 	}
 } 
 
+gdf_time_unit get_min_time_unit(gdf_time_unit unit1, gdf_time_unit unit2){
```
[the remark this answers is not in the document]
Comment (williamBlazing) [reply src/CalciteExpressionParsing.cpp]: i am guessing we will have to refactor this in the future, but this is the reality right now. I cant use what does not exist yet

File: src/CalciteExpressionParsing.cpp [hunk 10]
```diff
@@ -291,6 +280,84 @@ gdf_dtype get_output_type(gdf_dtype input_left_type, gdf_dtype input_right_type,
...
+		min_unit = TIME_UNIT_us;
+	} else if (unit1 == TIME_UNIT_ms || unit2 == TIME_UNIT_ms) {
+		min_unit = TIME_UNIT_ms;
+	} else if (unit1 == TIME_UNIT_s || unit2 == TIME_UNIT_s) {
+		min_unit = TIME_UNIT_s;
+	}
+	return min_unit;
+}
+
+bool get_common_type(gdf_dtype type1, gdf_dtype_extra_info info1, gdf_dtype type2, gdf_dtype_extra_info info2, gdf_dtype & type_out, gdf_dtype_extra_info & info_out){
```
[the remark this answers is not in the document]
Comment (williamBlazing) [reply src/CalciteExpressionParsing.cpp]: yes, this function allows for concatenating compatible types
